shortfin/python/array_binding.cc

```diff
@@ -495,117 +495,212 @@ void BindArray(py::module_ &m) {
+          .def("__repr__", &device_array::to_s)
+          .def("__str__", [](device_array &self) -> std::string {
+            auto contents = self.contents_to_s();
+            if (!contents) return "<<unmappable>>";
+            return *contents;
+          });
+
+  // Numpy Array Protocol, if numpy is available
+  py::object importlib = py::module_::import_("importlib");
+  if (!importlib.attr("util").attr("find_spec")("numpy").is_none()) {
+    device_array_class
+        .def("__array__",
+             [refs](device_array &self) {
+               py::object np = py::module_::import_("numpy");
+
+               auto &table = refs->element_type_array_type_code_table;
+               auto it = table.find(self.dtype());
+               if (it == table.end()) {
+                 throw std::invalid_argument(fmt::format(
+                     "Python array.array type code not known for dtype "
+                     "{}: Cannot access items",
+                     self.dtype().name()));
+               }
+
+               auto typeString = py::str();
+               if (it->first == DType::float16()) {
+                 // Special case to support float16.
+                 typeString = py::str(np.attr("dtype")(np.attr("float16")));
+               } else {
+                 typeString = py::str(it->second);
+               }
+
+               PyMapping *mapping;
+               py::object mapping_obj = CreateMappingObject(&mapping);
+               mapping->set_dtype(self.dtype());
+               self.storage().map_explicit(
+                   mapping->mapping(),
+                   static_cast<iree_hal_memory_access_bits_t>(
+                       IREE_HAL_MEMORY_ACCESS_READ));
+               auto items = mapping->GetItems(mapping_obj, refs.get());
+               return np.attr("array")(items, py::arg("dtype") = typeString);
+             })
+        .def_prop_ro("__array_interface__", [refs](device_array &self) {
+          py::object np = py::module_::import_("numpy");
+
+          py::dict interface;
+          interface["version"] = 3;
+          interface["strides"] = py::none();
+
+          auto shape = self.shape();
+          py::list shapeList;
+          for (size_t i = 0; i < shape.size(); ++i) {
+            shapeList.append(shape[i]);
+          }
+          py::tuple shape_tuple(py::cast(shapeList));
+          interface["shape"] = shape_tuple;
+
+          auto &table = refs->element_type_array_type_code_table;
+          auto it = table.find(self.dtype());
+          if (it == table.end()) {
+            throw std::invalid_argument(
+                fmt::format("Python array.array type code not known for dtype "
+                            "{}: Cannot access items",
+                            self.dtype().name()));
+          }
+
+          auto typeString = py::str();
+          if (it->first == DType::float16()) {
+            typeString = py::str(np.attr("dtype")(np.attr("float16")));
```

Comment:
This is a long way around to just get the literal "float16". That's basically never going to change or need this indirection. If you just change this to a string literal, you can avoid the entire numpy import and even remove the conditional definition.

Reply:
Replaced with a literal string, and removed the conditional definition. Also removed `__array__` function... not sure why I thought that was required, but everything works fine without it.